tests/unittest.py

```diff
@@ -229,13 +229,20 @@ def tearDown(orig: Callable[[], R]) -> R:
         #
         # The easiest way to do this would be to do a full GC after each test
         # run, but that is very expensive. Instead, we disable GC (above) for
-        # the duration of the test so that we only need to run a gen-0 GC, which
-        # is a lot quicker.
+        # the duration of the test and only run a gen-0 GC, which is a lot
+        # quicker. This doesn't clean up everything, since the TestCase
+        # instance still holds references to objects created during the test,
+        # such as HomeServers, so we do a full GC every so often.
 
         @around(self)
         def tearDown(orig: Callable[[], R]) -> R:
             ret = orig()
             gc.collect(0)
+            # Run a full GC every 50 gen-0 GCs.
+            gen0_stats = gc.get_stats()[0]
+            gen0_collections = gen0_stats["collections"]
+            if gen0_collections % 50 == 0:
+                gc.collect()
```

Comment:
Any idea why we're manually kicking the GC at all here? Is this to try and make tests more deterministic somehow? (Are we relying on `__del__` being called somewhere?)

Reply:
Failed `Deferred`s that haven't been awaited on / consumed will error out when they are GCed (and fail the currently running test somehow). If we let the garbage collector run at its own leisure, such failures would be logged against a random future test. At least I think that's what the original code is trying to trigger.